File: hkws_Development/Samples/C#/SimpleSamples/ConvertPixelType/ConvertPixelType.cs
using System;
using System.Collections.Generic;
using MvCamCtrl.NET;
using System.Runtime.InteropServices;
using System.IO;

namespace ConvertPixelType
{
    class ConvertPixelType
    {
        static bool IsMonoPixelFormat(MyCamera.MvGvspPixelType enType)
        {
            switch (enType)
            {
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono10:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono10_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono12:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono12_Packed:
                    return true;
                default:
                    return false;
            }
        }

        static bool IsColorPixelFormat(MyCamera.MvGvspPixelType enType)
        {
            switch (enType)
            {
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BGR8_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_YUV422_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_YUV422_YUYV_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR8:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG8:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB8:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG8:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB10:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB10_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG10:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG10_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG10:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG10_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR10:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR10_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB12:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB12_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG12:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG12_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG12:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG12_Packed:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR12:
                case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR12_Packed:
                    return true;
                default:
                    return false;
            }
        }

        static void Main(string[] args)
        {
            int nRet = MyCamera.MV_OK;
            MyCamera device = new MyCamera();
            IntPtr pBufForConvert = IntPtr.Zero;

            do
            {
                // ch:枚举设备 | en:Enum deivce
                MyCamera.MV_CC_DEVICE_INFO_LIST stDevList = new MyCamera.MV_CC_DEVICE_INFO_LIST();
                nRet = MyCamera.MV_CC_EnumDevices_NET(MyCamera.MV_GIGE_DEVICE | MyCamera.MV_USB_DEVICE, ref stDevList);
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Enum device failed:{0:x8}", nRet);
                    break;
                }
                Console.WriteLine("Enum device count :{0} \n", stDevList.nDeviceNum);
                if (0 == stDevList.nDeviceNum)
                {
                    break;
                }

                MyCamera.MV_CC_DEVICE_INFO stDevInfo;

                // ch:打印设备信息 | en:Print device info
                for (Int32 i = 0; i < stDevList.nDeviceNum; i++)
                {
                    stDevInfo = (MyCamera.MV_CC_DEVICE_INFO)Marshal.PtrToStructure(stDevList.pDeviceInfo[i], typeof(MyCamera.MV_CC_DEVICE_INFO));

                    if (MyCamera.MV_GIGE_DEVICE == stDevInfo.nTLayerType)
                    {
                        MyCamera.MV_GIGE_DEVICE_INFO stGigEDeviceInfo = (MyCamera.MV_GIGE_DEVICE_INFO)MyCamera.ByteToStruct(stDevInfo.SpecialInfo.stGigEInfo, typeof(MyCamera.MV_GIGE_DEVICE_INFO));
                        uint nIp1 = ((stGigEDeviceInfo.nCurrentIp & 0xff000000) >> 24);
                        uint nIp2 = ((stGigEDeviceInfo.nCurrentIp & 0x00ff0000) >> 16);
                        uint nIp3 = ((stGigEDeviceInfo.nCurrentIp & 0x0000ff00) >> 8);
                        uint nIp4 = (stGigEDeviceInfo.nCurrentIp & 0x000000ff);
                        Console.WriteLine("[device " + i.ToString() + "]:");
                        Console.WriteLine("DevIP:" + nIp1 + "." + nIp2 + "." + nIp3 + "." + nIp4);
                        Console.WriteLine("UserDefineName:" + stGigEDeviceInfo.chUserDefinedName + "\n");
                    }
                    else if (MyCamera.MV_USB_DEVICE == stDevInfo.nTLayerType)
                    {
                        MyCamera.MV_USB3_DEVICE_INFO stUsb3DeviceInfo = (MyCamera.MV_USB3_DEVICE_INFO)MyCamera.ByteToStruct(stDevInfo.SpecialInfo.stUsb3VInfo, typeof(MyCamera.MV_USB3_DEVICE_INFO));
                        Console.WriteLine("[device " + i.ToString() + "]:");
                        Console.WriteLine("SerialNumber:" + stUsb3DeviceInfo.chSerialNumber);
                        Console.WriteLine("UserDefineName:" + stUsb3DeviceInfo.chUserDefinedName + "\n");
                    }
                }

                Int32 nDevIndex = 0;
                Console.Write("Please input index(0-{0:d}):", stDevList.nDeviceNum - 1);
                try
                {
                    nDevIndex = Convert.ToInt32(Console.ReadLine());
                }
                catch
                {
                    Console.Write("Invalid Input!\n");
                    break;
                }

                if (nDevIndex > stDevList.nDeviceNum - 1 || nDevIndex < 0)
                {
                    Console.Write("Input Error!\n");
                    break;
                }
                stDevInfo = (MyCamera.MV_CC_DEVICE_INFO)Marshal.PtrToStructure(stDevList.pDeviceInfo[nDevIndex], typeof(MyCamera.MV_CC_DEVICE_INFO));

                // ch:创建设备 | en: Create device
                nRet = device.MV_CC_CreateDevice_NET(ref stDevInfo);
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Create device failed:{0:x8}", nRet);
                    break;
                }

                // ch:打开设备 | en:Open device
                nRet = device.MV_CC_OpenDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Open device failed:{0:x8}", nRet);
                    break;
                }

                // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
                if (stDevInfo.nTLayerType == MyCamera.MV_GIGE_DEVICE)
                {
                    int nPacketSize = device.MV_CC_GetOptimalPacketSize_NET();
                    if (nPacketSize > 0)
                    {
                        nRet = device.MV_CC_SetIntValue_NET("GevSCPSPacketSize", (uint)nPacketSize);
                        if (nRet != MyCamera.MV_OK)
                        {
                            Console.WriteLine("Warning: Set Packet Size failed {0:x8}", nRet);
                        }
                    }
                    else
                    {
                        Console.WriteLine("Warning: Get Packet Size failed {0:x8}", nPacketSize);
                    }
                }

                // ch:设置触发模式为off || en:set trigger mode as off
                nRet = device.MV_CC_SetEnumValue_NET("TriggerMode", 0);
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Set TriggerMode failed:{0:x8}", nRet);
                    break;
                }

                // ch:开启抓图 || en: start grab image
                nRet = device.MV_CC_StartGrabbing_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Start grabbing failed:{0:x8}", nRet);
                    break;
                }

                MyCamera.MV_FRAME_OUT stFrameOut = new MyCamera.MV_FRAME_OUT();

                nRet = device.MV_CC_GetImageBuffer_NET(ref stFrameOut, 1000);
                // 获取一帧图像
                if (MyCamera.MV_OK == nRet)
                {
                    Console.WriteLine("Get One Frame:" + "Width[" + Convert.ToString(stFrameOut.stFrameInfo.nWidth) + "] , Height["
                        + Convert.ToString(stFrameOut.stFrameInfo.nHeight) + "] , FrameNum[" + Convert.ToString(stFrameOut.stFrameInfo.nFrameNum) + "]");

                    MyCamera.MvGvspPixelType enType = MyCamera.MvGvspPixelType.PixelType_Gvsp_Undefined;
                    uint nChannelNum = 0;
                    if (IsColorPixelFormat(stFrameOut.stFrameInfo.enPixelType))
                    {
                        enType = MyCamera.MvGvspPixelType.PixelType_Gvsp_RGB8_Packed;
                        nChannelNum = 3;
                    }
                    else if (IsMonoPixelFormat(stFrameOut.stFrameInfo.enPixelType))
                    {
                        enType = MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono8;
                        nChannelNum = 1;
                    }
                    else
                    {
                        Console.WriteLine("Don't need to convert!");
                    }

                    if (enType != MyCamera.MvGvspPixelType.PixelType_Gvsp_Undefined)
                    {
                        if (pBufForConvert == IntPtr.Zero)
                        {
                            pBufForConvert = Marshal.AllocHGlobal((int)(stFrameOut.stFrameInfo.nWidth * stFrameOut.stFrameInfo.nHeight * nChannelNum));
                        }
                        MyCamera.MV_PIXEL_CONVERT_PARAM stConvertPixelParam = new MyCamera.MV_PIXEL_CONVERT_PARAM();
                        stConvertPixelParam.nWidth = stFrameOut.stFrameInfo.nWidth;
                        stConvertPixelParam.nHeight = stFrameOut.stFrameInfo.nHeight;
                        stConvertPixelParam.pSrcData = stFrameOut.pBufAddr;
                        stConvertPixelParam.nSrcDataLen = stFrameOut.stFrameInfo.nFrameLen;
                        stConvertPixelParam.enSrcPixelType = stFrameOut.stFrameInfo.enPixelType;
                        stConvertPixelParam.enDstPixelType = enType;
                        stConvertPixelParam.pDstBuffer = pBufForConvert;
                        stConvertPixelParam.nDstBufferSize = (uint)(stFrameOut.stFrameInfo.nWidth * stFrameOut.stFrameInfo.nHeight * nChannelNum);

                        nRet = device.MV_CC_ConvertPixelType_NET(ref stConvertPixelParam);
                        if (MyCamera.MV_OK != nRet)
                        {
                            Console.WriteLine("Convert pixel type Failed:{0:x8}", nRet);
                            break;
                        }

                        // ch:将图像数据保存到本地文件 | en:Save image data to local file
                        byte[] data = new byte[stConvertPixelParam.nDstLen];
                        Marshal.Copy(stConvertPixelParam.pDstBuffer, data, 0, (int)stConvertPixelParam.nDstLen);
                        FileStream pFile = null;
                        try
                        {
                            pFile = new FileStream("AfterConvert.raw", FileMode.Create);
                            pFile.Write(data, 0, data.Length);
                        }
                        catch
                        {
                            Console.WriteLine("保存失败");
                        }
                        finally
                        {
                            pFile.Close();
                        }
                    }
                    device.MV_CC_FreeImageBuffer_NET(ref stFrameOut);
                }
                else
                {
                    Console.WriteLine("Get Image failed:{0:x8}", nRet);
                }

                // ch:停止抓图 | en:Stop grabbing
                nRet = device.MV_CC_StopGrabbing_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Stop grabbing failed:{0:x8}", nRet);
                    break;
                }

                // ch:关闭设备 | en:Close device
                nRet = device.MV_CC_CloseDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Close device failed:{0:x8}", nRet);
                    break;
                }

                // ch:销毁设备 | en:Destroy device
                nRet = device.MV_CC_DestroyDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Destroy device failed:{0:x8}", nRet);
                    break;
                }
            } while (false);

            if (pBufForConvert != IntPtr.Zero)
            {
                Marshal.FreeHGlobal(pBufForConvert);
            }

            if (MyCamera.MV_OK != nRet)
            {
                // ch:销毁设备 | en:Destroy device
                nRet = device.MV_CC_DestroyDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Destroy device failed:{0:x8}", nRet);
                }
            }

            Console.WriteLine("Press enter to exit");
            Console.ReadKey();
        }
    }
}

File: hkws_Development/Samples/VB/VBDotNet/SimpleSameples/ConvertPixelType/ConvertPixelType.vb
Imports System.Runtime.InteropServices
Imports System.Threading.Thread
Imports System.Net.IPAddress
Imports MvCamCtrl.NET

Module ConvertPixelType

    Function IsMonoPixelFormat(ByVal enType As MyCamera.MvGvspPixelType)
        Select Case enType
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono10
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono10_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono12
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono12_Packed
                Return True
        End Select
        Return False
    End Function

    Function IsColorPixelFormat(ByVal enType As MyCamera.MvGvspPixelType)
        Select Case enType
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BGR8_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_YUV422_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_YUV422_YUYV_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR8
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG8
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB8
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG8
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB10
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB10_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG10
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG10_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG10
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG10_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR10
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR10_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB12
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGB12_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG12
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerBG12_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG12
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerRG12_Packed
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR12
                Return True
            Case MyCamera.MvGvspPixelType.PixelType_Gvsp_BayerGR12_Packed
                Return True
        End Select
        Return False
    End Function

    Sub Main()
        Dim dev As MyCamera = New MyCamera
        Dim nRet As Int32 = MyCamera.MV_OK
        Dim pBufForConvert As IntPtr = IntPtr.Zero

        Do While (True)
            Dim stDeviceInfoList As MyCamera.MV_CC_DEVICE_INFO_LIST = New MyCamera.MV_CC_DEVICE_INFO_LIST

            ' ch:枚举设备 | en:Enum device
            nRet = MyCamera.MV_CC_EnumDevices_NET((MyCamera.MV_GIGE_DEVICE Or MyCamera.MV_USB_DEVICE), stDeviceInfoList)
            If MyCamera.MV_OK <> nRet Then
                Console.WriteLine("Enum Device failed:{0:x8}", nRet)
                Exit Do
            End If

            If (0 = stDeviceInfoList.nDeviceNum) Then
                Console.WriteLine("No Find Gige | Usb Device !")
                Exit Do
            End If

            ' ch:打印设备信息 | en:Print device info
            Dim i As Int32
            For i = 0 To stDeviceInfoList.nDeviceNum - 1
                Dim stDeviceInfo As MyCamera.MV_CC_DEVICE_INFO = New MyCamera.MV_CC_DEVICE_INFO
                stDeviceInfo = CType(Marshal.PtrToStructure(stDeviceInfoList.pDeviceInfo(i), GetType(MyCamera.MV_CC_DEVICE_INFO)), MyCamera.MV_CC_DEVICE_INFO)
                If (MyCamera.MV_GIGE_DEVICE = stDeviceInfo.nTLayerType) Then
                    Dim stGigeInfoPtr As IntPtr = Marshal.AllocHGlobal(216)
                    Marshal.Copy(stDeviceInfo.SpecialInfo.stGigEInfo, 0, stGigeInfoPtr, 216)
                    Dim stGigeInfo As MyCamera.MV_GIGE_DEVICE_INFO
                    stGigeInfo = CType(Marshal.PtrToStructure(stGigeInfoPtr, GetType(MyCamera.MV_GIGE_DEVICE_INFO)), MyCamera.MV_GIGE_DEVICE_INFO)
                    Dim nIpByte1 As UInt32 = (stGigeInfo.nCurrentIp And &HFF000000) >> 24
                    Dim nIpByte2 As UInt32 = (stGigeInfo.nCurrentIp And &HFF0000) >> 16
                    Dim nIpByte3 As UInt32 = (stGigeInfo.nCurrentIp And &HFF00) >> 8
                    Dim nIpByte4 As UInt32 = (stGigeInfo.nCurrentIp And &HFF)

                    Console.WriteLine("[Dev " + Convert.ToString(i) + "]:")
                    Console.WriteLine("DevIP:" + nIpByte1.ToString() + "." + nIpByte2.ToString() + "." + nIpByte3.ToString() + "." + nIpByte4.ToString())
                    Console.WriteLine("UserDefinedName:" + stGigeInfo.chUserDefinedName)
                    Console.WriteLine("")
                Else
                    Dim stUsbInfoPtr As IntPtr = Marshal.AllocHGlobal(540)
                    Marshal.Copy(stDeviceInfo.SpecialInfo.stUsb3VInfo, 0, stUsbInfoPtr, 540)
                    Dim stUsbInfo As MyCamera.MV_USB3_DEVICE_INFO
                    stUsbInfo = CType(Marshal.PtrToStructure(stUsbInfoPtr, GetType(MyCamera.MV_USB3_DEVICE_INFO)), MyCamera.MV_USB3_DEVICE_INFO)

                    Console.WriteLine("[Dev " + Convert.ToString(i) + "]:")
                    Console.WriteLine("SerialNumber:" + stUsbInfo.chSerialNumber)
                    Console.WriteLine("UserDefinedName:" + stUsbInfo.chUserDefinedName)
                    Console.WriteLine("")
                End If
            Next

            Console.Write("Please input index(0-{0:d}):", stDeviceInfoList.nDeviceNum - 1)
            Dim nIndex As Int32
            Try
                nIndex = Console.ReadLine()
            Catch ex As Exception
                Console.WriteLine("Invalid input!")
                Console.WriteLine("push enter to exit")
                System.Console.ReadLine()
                End
            End Try

            If nIndex > stDeviceInfoList.nDeviceNum - 1 Then
                Console.WriteLine("Invalid input!")
                Console.WriteLine("push enter to exit")
                System.Console.ReadLine()
                End
            End If

            If nIndex < 0 Then
                Console.WriteLine("Invalid input!")
                Console.WriteLine("push enter to exit")
                System.Console.ReadLine()
                End
            End If


            Dim stdevInfo As MyCamera.MV_CC_DEVICE_INFO
            stdevInfo = CType(Marshal.PtrToStructure(stDeviceInfoList.pDeviceInfo(nIndex), GetType(MyCamera.MV_CC_DEVICE_INFO)), MyCamera.MV_CC_DEVICE_INFO)

            ' ch:创建句柄 | en:Create handle
            nRet = dev.MV_CC_CreateDevice_NET(stdevInfo)
            If 0 <> nRet Then
                Console.WriteLine("Create device failed:{0:x8}", nRet)
                Exit Do
            End If

            ' ch:打开相机 | en:Open device
            nRet = dev.MV_CC_OpenDevice_NET()
            If 0 <> nRet Then
                Console.WriteLine("Open device failed:{0:x8}", nRet)
                Exit Do
            End If

            ' ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
            If stdevInfo.nTLayerType = MyCamera.MV_GIGE_DEVICE Then
                Dim nPacketSize As Int32
                nPacketSize = dev.MV_CC_GetOptimalPacketSize_NET()
                If nPacketSize > 0 Then
                    nRet = dev.MV_CC_SetIntValue_NET("GevSCPSPacketSize", nPacketSize)
                    If 0 <> nRet Then
                        Console.WriteLine("Warning: Set Packet Size failed:{0:x8}", nRet)
                    End If
                Else
                    Console.WriteLine("Warning: Get Packet Size failed:{0:x8}", nPacketSize)
                End If
            End If

            ' ch:开启取流 | en:Start grabbing
            nRet = dev.MV_CC_StartGrabbing_NET()
            If 0 <> nRet Then
                Console.WriteLine("Start grabbing fauled:{0:x8}", nRet)
                Exit Do
            End If

            Dim stFrameOut As MyCamera.MV_FRAME_OUT = New MyCamera.MV_FRAME_OUT

            ' ch:抓取图像 | en:Get image
            nRet = dev.MV_CC_GetImageBuffer_NET(stFrameOut, 1000)
            If MyCamera.MV_OK = nRet Then
                Console.WriteLine("Width:" + Convert.ToString(stFrameOut.stFrameInfo.nWidth) + " Height:" + Convert.ToString(stFrameOut.stFrameInfo.nHeight) + " FrameNum:" + Convert.ToString(stFrameOut.stFrameInfo.nFrameNum))

                Dim enType As MyCamera.MvGvspPixelType = MyCamera.MvGvspPixelType.PixelType_Gvsp_Undefined
                Dim nChannelNum As Int32 = 0
                If IsColorPixelFormat(stFrameOut.stFrameInfo.enPixelType) = True Then
                    enType = MyCamera.MvGvspPixelType.PixelType_Gvsp_RGB8_Packed
                    nChannelNum = 3
                ElseIf IsMonoPixelFormat(stFrameOut.stFrameInfo.enPixelType) = True Then
                    enType = MyCamera.MvGvspPixelType.PixelType_Gvsp_Mono8
                    nChannelNum = 1
                Else
                    Console.WriteLine("Don't need to convert!")
                End If

                Dim nWidth As Int32 = stFrameOut.stFrameInfo.nWidth
                Dim nHeight As Int32 = stFrameOut.stFrameInfo.nHeight

                If MyCamera.MvGvspPixelType.PixelType_Gvsp_Undefined <> enType Then
                    If pBufForConvert = IntPtr.Zero Then
                        
                        Dim nBufForConvertSzie As Integer = nWidth * nHeight * nChannelNum
                        pBufForConvert = Marshal.AllocHGlobal(nBufForConvertSzie)
                    End If
                    Dim stConvertParam As MyCamera.MV_PIXEL_CONVERT_PARAM = New MyCamera.MV_PIXEL_CONVERT_PARAM()
                    stConvertParam.nHeight = stFrameOut.stFrameInfo.nHeight
                    stConvertParam.nWidth = stFrameOut.stFrameInfo.nWidth
                    stConvertParam.pSrcData = stFrameOut.pBufAddr
                    stConvertParam.nSrcDataLen = stFrameOut.stFrameInfo.nFrameLen
                    stConvertParam.enSrcPixelType = stFrameOut.stFrameInfo.enPixelType
                    stConvertParam.enDstPixelType = enType
                    stConvertParam.pDstBuffer = pBufForConvert
                    stConvertParam.nDstBufferSize = nWidth * nHeight * nChannelNum

                    nRet = dev.MV_CC_ConvertPixelType_NET(stConvertParam)
                    If MyCamera.MV_OK <> nRet Then
                        Console.WriteLine("Convert failed:{0:x8}", nRet)
                        Exit Do
                    End If

                    ' ch:将图像数据保存到本地文件 | en:Save image data to local file
                    Dim m_byteImageBuffer(stConvertParam.nDstLen - 1) As Byte
                    Marshal.Copy(pBufForConvert, m_byteImageBuffer, 0, stConvertParam.nDstLen)
                    Try
                        My.Computer.FileSystem.WriteAllBytes("AfterConvert.raw", m_byteImageBuffer, False)
                    Catch ex As Exception
                    End Try
                End If
                dev.MV_CC_FreeImageBuffer_NET(stFrameOut)
            Else
                Console.WriteLine("Get Image failed:{0:x8}", nRet)
                Exit Do
            End If

            ' ch:停止取流 | en:Stop grabbing
            nRet = dev.MV_CC_StopGrabbing_NET()
            If 0 <> nRet Then
                Console.WriteLine("Stop Grabbing failed:{0:x8}", nRet)
                Exit Do
            End If

            ' ch:关闭相机 | en:Close device
            nRet = dev.MV_CC_CloseDevice_NET()
            If 0 <> nRet Then
                Console.WriteLine("Open device failed:{0:x8}", nRet)
                Exit Do
            End If

            ' ch:销毁句柄 | en:Destroy handle
            nRet = dev.MV_CC_DestroyDevice_NET()
            If 0 <> nRet Then
                Console.WriteLine("Destroy device failed:{0:x8}", nRet)
                Exit Do
            End If

            Exit Do
        Loop

        If IntPtr.Zero <> pBufForConvert Then
            Marshal.FreeHGlobal(pBufForConvert)
        End If

        If 0 <> nRet Then
            ' ch:销毁句柄 | en:Destroy handle
            nRet = dev.MV_CC_DestroyDevice_NET()
            If 0 <> nRet Then
                Console.WriteLine("Destroy device failed:{0:x8}", nRet)
            End If
        End If


        Console.WriteLine("Press enter to exit")
        System.Console.ReadLine()
    End Sub

End Module
